crates/ursa-network/src/origin.rs

```diff
@@ -0,0 +1,178 @@
+use std::collections::VecDeque;
+use std::fmt::{Debug, Display};
+use std::slice::Iter;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use anyhow::{anyhow, Result};
+use cid::Cid;
+use db::Store;
+use fnv::FnvHashMap;
+use futures_util::FutureExt;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::load_car;
+use libp2p::{
+    core::connection::ConnectionId,
+    swarm::{
+        behaviour::FromSwarm, dummy::ConnectionHandler, NetworkBehaviour, NetworkBehaviourAction,
+        PollParameters,
+    },
+    PeerId,
+};
+use surf::Client;
+use tokio::sync::oneshot::{channel, Receiver};
+use tracing::error;
+use void::Void;
+
+use ursa_store::UrsaStore;
+
+use crate::OriginConfig;
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum Origin {
+    Ipfs,
+    Arweave,
+    Filecoin,
+}
+
+impl Origin {
+    pub fn iter() -> Iter<'static, Origin> {
+        [Origin::Ipfs].iter()
+    }
+
+    pub async fn get<G: Display, C: Display>(
+        &self,
+        client: &Client,
+        gateway: G,
+        cid: C,
+    ) -> Result<Vec<u8>> {
+        let url = match self {
+            Origin::Ipfs => format!("https://{gateway}/ipfs/{cid}"),
+            _ => unimplemented!(),
+        };
+
+        client
+            .get(url)
+            .header("Accept", "application/vnd.ipld.car")
+            .recv_bytes()
+            .await
+            .map_err(|e| anyhow!(e))
+    }
+}
+
+pub type QueryId = u64;
+
+pub struct PendingQuery {
+    pub cid: Cid,
+    pub receiver: Receiver<Result<()>>,
+}
+
+#[derive(Debug)]
+pub enum OriginEvent {
+    QueryCompleted(QueryId, Cid, Result<()>),
+}
+
+pub struct OriginBehavior<S> {
+    store: Arc<UrsaStore<S>>,
+    query_count: QueryId,
+    pending: FnvHashMap<QueryId, PendingQuery>,
+    client: Arc<Client>,
+    config: OriginConfig,
+    event_queue: VecDeque<OriginEvent>,
+}
+
+impl<S> OriginBehavior<S>
```

Comment:
Why do we need a behaviour to fetch data from the origin if we're just going to use `surf::Client`, wouldn't it be more simple to just make that call in `check_blockstore`?

Reply:
The goal is to share successful responses for concurrent requests for cids from both bitswap and origin

Also, as we add more origins the requirements to get from them will differ and extend further than simple http calls

And finally to note I may switch surf client -> hyper client here, lmk if you have thoughts on it

Reply:
Oh I see. You don't need a behaviour to communicate to Bitswap about an event though, you can use a channel for that. 

> Also, as we add more implementations for origins, the requirements to get from them will differ and extend further than just simple http calls

I recommend that we get a better idea of requirements before writing code for it. You can avoid a lot of work if you know what you need.

I like the idea of concurrent requests but we should be cautious of optimizing too early in development before having some data and being able to measure the trade off such as bandwidth and latency.

Reply:
https://github.com/fleek-network/ursa/pull/278/commits/11e9610901aad513e46cbede8f05907d30ba1363